src/graph/on_disk_graph.cpp

```diff
@@ -34,38 +45,52 @@ static std::unique_ptr<RelTableScanState> getRelScanState(MemoryManager& mm, con
     scanState->nodeIDVector = srcVector;
     scanState->outputVectors.push_back(dstVector);
     scanState->outputVectors.push_back(relIDVector);
+    for (auto& property : properties) {
+        auto pos = DataPos(schema.getExpressionPos(*property));
+        auto vector = resultSet.getValueVector(pos).get();
+        scanState->outputVectors.push_back(vector);
+    }
     scanState->outState = dstVector->state.get();
     return scanState;
 }
 
-OnDiskGraphScanState::OnDiskGraphScanState(ClientContext* context, RelTable& table,
-    ValueVector* srcNodeIDVector, ValueVector* dstNodeIDVector, ValueVector* relIDVector)
-    : fwdIterator{context, &table,
-          getRelScanState(*context->getMemoryManager(), table, RelDataDirection::FWD,
-              srcNodeIDVector, dstNodeIDVector, relIDVector)},
-      bwdIterator{context, &table,
-          getRelScanState(*context->getMemoryManager(), table, RelDataDirection::BWD,
-              srcNodeIDVector, dstNodeIDVector, relIDVector)} {}
-
-OnDiskGraphScanStates::OnDiskGraphScanStates(ClientContext* context, std::span<RelTable*> tables)
-    : iteratorIndex{0}, direction{RelDataDirection::FWD} {
-    scanStates.reserve(tables.size());
-    srcNodeIDVectorState = DataChunkState::getSingleValueDataChunkState();
-    dstNodeIDVectorState = std::make_shared<DataChunkState>();
+OnDiskGraphScanStates::OnDiskGraphScanStates(ClientContext* context, std::span<RelTable*> tables,
+    const GraphEntry& graphEntry)
+    : iteratorIndex{0}, direction{RelDataDirection::INVALID} {
+    auto schema = Schema();
```

Comment:
I don't have a very concrete suggestion here but I'm a bit worried that some of this logic of how we map an expression to scan operators must be also implemented in other parts of the mapper (e.g., when we compile scan + filter operators) and is being duplicated here. I wonder if there is some way we can refactor this logic. For example some function that takes in predicate expression and figures out the properties that needs to be scanned from the expression and constructs a predicateEvaluator and update a ResultSet. And maybe it's used here and in the mapper?

Reply:
I only make a minimal change. I understand your comment. Though the problem needs a relatively bigger change which is beyond the scope of this PR. I'll make a following up PR soon to resolve this.